README.rst
.. image:: https://travis-ci.org/kitconcept/pytest-solr.svg?branch=master
    :target: https://travis-ci.org/kitconcept/pytest-solr

.. image:: https://img.shields.io/pypi/v/pytest-solr.svg
    :target: https://pypi.python.org/pypi/pytest-solr/
    :alt: Latest Version

.. image:: https://img.shields.io/pypi/status/pytest-solr.svg
    :target: https://pypi.python.org/pypi/pytest-solr/
    :alt: Egg Status

.. image:: https://img.shields.io/pypi/l/pytest-solr.svg
    :target: https://pypi.python.org/pypi/pytest-solr/
    :alt: License

Solr process and client fixtures for py.test.

.. image:: https://raw.githubusercontent.com/kitconcept/pytest-solr/master/kitconcept.png
   :alt: kitconcept
   :target: https://kitconcept.com/

Introduction
------------

pytest-solr is a pytest plugin for the Apache Solr search server.
It provides three pytest factories:

solr_process:
  For starting and stopping the Solr server. This is session scoped.

solr_core:
  For loading and unloading a Solr core configuration. This is module scoped.

solr:
  For connecting to a Solr server during a test. This is function scoped.


Solr Process
^^^^^^^^^^^^

The solr_process factory starts and stops a the Solr process.
An existing Solr executable is required for this.

'executable':
  path to the Solr executable. Default value is 'downloads/solr-<SOLR_VERSION>/bin/solr'
'host':
  hostname where Solr runs. Default value is 'localhost'.
'port':
  port Solr uses. Default is value is '18983'.
'core':
  Solr core that is used. Default value is 'solr'.
'timeout':
  timeout to wait for Solr to start. Default value is '60' (seconds).

Example::

  from pytest_solr.factories import solr_process

  solr_process = solr_process(
    executable='solr-6.5.0/bin/solr',
    host='localhost',
    port=8983,
    core='default',
    version='6.5.0',
    timeout=60
  )


Solr Core
^^^^^^^^^

The solr_core factory adds and removes a Solr core configuration.
It expects two parameters, the Solr Process fixture name and the Solr core name.

'solr_process_fixture_name':
  String with the name of the Solr Process. This is a required parameter.
'solr_core_name':
  String with the name of the Solr core. Default value is 'default'.

Example::

  from pytest_solr.factories import solr_core

  my_solr_core = solr_core('solr_process', 'my_solr_core')


Solr
^^^^

The Solr factory connects to Solr via pysolr.
It expects a single parameter, the Solr core fixture name.

'solr_core_fixture_name':
  String with the name of the Solr core. This is a required parameter.

Example::

  # -*- coding: utf-8 -*-
  from pytest_solr.factories import solr_core
  from pytest_solr.factories import solr

  minimal = solr_core('solr_process', 'minimal')
  solr = solr('minimal')

  def test_exact_term_match(solr):
      solr.add([{'id': '1', 'title': 'bananas'}])
      assert 1 == solr.search('title:bananas').hits

The solr fixture can then be injected into the test function and used to add documents to solr or search for terms.

See the `pysolr documentation <https://github.com/django-haystack/pysolr>`_. for more details.


Installation
------------

Install pytest-solr with pip::

  $ pip install pytest-solr


Usage
-----

Create a solr core with the name 'minimal' and inject the use the solr factory into a test function to use it::

    # -*- coding: utf-8 -*-
    from pytest_solr.factories import solr_core
    from pytest_solr.factories import solr

    minimal = solr_core('solr_process', 'minimal')
    solr = solr('minimal')


    def test_exact_term_match(solr):
        solr.add([{'id': '1', 'title': 'bananas'}])
        assert 1 == solr.search('title:bananas').hits

License
-------

Copyright kitconcept GmbH.

Distributed under the terms of the MIT license, pytest-solr is free and Open Source software.


Contribute
----------

- `Source code at Github <https://github.com/kitconcept/pytest-solr>`_
- `Issue tracker at Github <https://github.com/kitconcept/pytest-solr/issues>`_


Support
-------

If you are having issues, `please let us know <https://github.com/kitconcept/pytest-solr/issues>`_. If you require professional support feel free to contact us at `info@kitconcept.com. <mailto:info@kitconcept.com>`_


CHANGES.rst
1.0.1 (unreleased)
------------------

- Nothing changed yet.


1.0.0 (2020-05-11)
------------------

- Set Solr to always_commit=True.
  [timo]

- Black.
  [timo]

- Make Solr 7.7.3 the new default.
  [timo]


1.0a1 (2017-06-22)
------------------

- Initial release.
  [timo]



tests/test_solr.py
# -*- coding: utf-8 -*-
from pytest_solr.factories import solr_core
from pytest_solr.factories import solr


def test_solr_process(solr_process):
    assert solr_process.get("process").poll() is None


def test_solr_core(solr_core):
    assert solr_core


# def test_solr(solr):
#     assert 0 == solr.search('bananas').hits


# solr_core_custom = solr_core('solr_process', 'minimal')
# solr_custom = solr('solr_core_custom')


# def test_solr_core_custom(solr_custom):
#     solr_custom.add([{'id': '1', 'title': 'bananas'}])
#     assert 1 == solr_custom.search('title:bananas').hits


tests/plone/test_plone.py
# -*- coding: utf-8 -*-
from pytest_solr.factories import solr_core
from pytest_solr.factories import solr

plone = solr_core('solr_process', 'plone')
solr = solr('plone')


def test_title(solr):
    solr.add([{'id': '1', 'Title': 'bananas'}])
    assert 1 == solr.search('Title:bananas').hits


def test_description(solr):
    solr.add([{'id': '1', 'Description': 'bananas'}])
    assert 1 == solr.search('Description:bananas').hits


def test_search_ignores_lowercase(solr):
    solr.add([{'id': '1', 'Title': 'Bananas'}])
    assert 1 == solr.search('Title:bananas').hits


def test_search_ignores_uppercase(solr):
    solr.add([{'id': '1', 'Title': 'bananas'}])
    assert 1 == solr.search('Title:Bananas').hits


def test_title_indexed_in_searchable_text(solr):
    solr.add([{'id': '1', 'Title': 'bananas'}])
    assert 1 == solr.search('default:bananas').hits


def test_description_indexed_in_searchable_text(solr):
    solr.add([{'id': '1', 'Description': 'bananas'}])
    assert 1 == solr.search('default:bananas').hits


def test_creator_indexed_in_searchable_text(solr):
    solr.add([{'id': '1', 'Creator': 'bananas'}])
    assert 1 == solr.search('default:bananas').hits


tests/default/test_default.py
# -*- coding: utf-8 -*-
from pytest_solr.factories import solr_core
from pytest_solr.factories import solr

minimal = solr_core("solr_process", "minimal")
solr = solr("minimal")


def test_exact_term_match(solr):
    solr.add([{"id": "1", "title": "bananas"}])
    assert 1 == solr.search("title:bananas").hits


src/pytest_solr/factories.py
import os
import pysolr
import pytest
import signal
import subprocess


def solr_process(
    executable=None,
    host="localhost",
    port=18983,
    core="solr",
    version="7.7.3",
    timeout=60,
):
    @pytest.fixture(scope="session")
    def solr_process_fixture(request):
        solr = {
            "host": host,
            "port": port,
        }

        if not os.environ.get("SOLR_VERSION"):
            solr["version"] = version
        else:
            solr["version"] = os.environ["SOLR_VERSION"]

        if not executable:
            local_executable = "downloads/solr-{}/bin/solr".format(solr["version"])

        solr["directory"] = "downloads/solr-{}".format(solr["version"])
        solr["bin"] = "{}/bin/solr".format(solr["directory"])

        devnull = open("/dev/null", "w")
        solr["process"] = subprocess.Popen(
            "{} -f -p {}".format(executable or local_executable, port),
            stdout=devnull,
            stderr=devnull,
            shell=True,
            preexec_fn=os.setsid,
        )

        yield solr

        # tear down
        os.killpg(solr["process"].pid, signal.SIGTERM)

    return solr_process_fixture


def solr_core(solr_process_fixture_name, solr_core_name="default"):
    @pytest.fixture(scope="module")
    def solr_core_fixture(request):
        process = request.getfixturevalue(solr_process_fixture_name)
        solr_executable = process.get("bin")
        solr_core_directory = "tests/{}".format(solr_core_name)
        solr_port = str(process.get("port"))

        def create_solr_colr():
            try:
                subprocess.check_output(
                    [
                        solr_executable,
                        "create_core",
                        "-p",
                        solr_port,
                        "-c",
                        solr_core_name,
                        "-d",
                        solr_core_directory,
                    ],
                )
            except subprocess.CalledProcessError as e:
                print(
                    "Creating Solr Core failed. "
                    + 'Command "{}" failed with the following error: {}'.format(
                        " ".join(e.cmd), e.output
                    )
                )

        def drop_solr_core():
            subprocess.check_output(
                [solr_executable, "delete", "-p", solr_port, "-c", solr_core_name,],
            )

        try:
            create_solr_colr()
        except subprocess.CalledProcessError:
            drop_solr_core()
            create_solr_colr()

        yield process

        drop_solr_core()

    return solr_core_fixture


def solr(solr_core_fixture_name):
    @pytest.fixture(scope="function")
    def solr_fixture(request):
        process = request.getfixturevalue(solr_core_fixture_name)

        client = pysolr.Solr(
            "http://{0!s}:{1!s}/solr/{2!s}".format(
                process.get("host"), process.get("port"), solr_core_fixture_name
            ),
            always_commit=True,
        )

        yield client

        client.delete(q="*:*")

    return solr_fixture
